mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/SyncServiceImpl.kt

```diff
@@ -63,45 +77,134 @@ class SyncServiceImpl(userNotifier: INotifier) : ISyncService {
         val languageRepository = registerLanguages(targetProject)
-        return runBlocking(dispatcher) {
-            BranchRegistry.setBranch(
-                client,
-                branchReference,
-                languageRepository,
-                targetProject,
-                CoroutineScope(dispatcher),
-            )
-        }
+        val model = BranchRegistry.setReplicatedModel(
+            client,
+            branchReference,
+            languageRepository,
+            targetProject,
+            ReplicatedModelInitContext(CoroutineScope(dispatcher), initialVersion),
+        )
+        logger.info { "Connected to branch $branchReference with initial version $initialVersion" }
+        return model
     }
 
     /**
-     * WARNING: this is a long-running blocking call.
-     * Do not call this method from the main / EDT Thread, otherwise it will not be able to write to MPS!!!
+     * WARNING:
+     * 1. This is a long-running blocking call.
+     * 2. Do not call this method from the main / EDT Thread, otherwise it will not be able to write to MPS!!!
      */
     override fun bindModuleFromServer(
         client: ModelClientV2,
         branchReference: BranchReference,
-        moduleId: String,
+        module: ModuleIdWithName,
     ): Iterable<IBinding> {
+        val moduleName = module.name
+        logger.info { "Binding Module '$moduleName' from the server ($branchReference)." }
+
         val targetProject = mpsProjectInjector.activeMpsProject!!
         val languageRepository = registerLanguages(targetProject)
 
         // fetch replicated model and branch content
         val branch = connectToBranch(client, branchReference)
 
         // transform the modules and models
-        val bindings = ITreeToSTreeTransformer(branch, languageRepository).transform(moduleId)
+        val bindings = ITreeToSTreeTransformer(branch, languageRepository).transform(module.id)
+
+        notifyUserAboutBindings(bindings, moduleName)
+
+        return bindings
+    }
+
+    /**
+     * WARNING: this is a long-running blocking call.
+     */
+    override fun rebindModules(
+        client: ModelClientV2,
+        branchReference: BranchReference,
+        initialVersion: CLVersion,
+        modules: Iterable<AbstractModule>,
+    ): Iterable<IBinding>? {
+        if (!modules.iterator().hasNext()) {
+            val message =
+                "The list is restorable Modules is empty, therefore no Module- or Model Binding is restored for them."
+            notifierInjector.notifyAndLogWarning(message, logger)
+            return null
+        }
+
+        // connect to modelix and fetch the initial version
+        val replicatedModel = setReplicatedModel(client, branchReference, initialVersion)
+
+        // recreate the mapping between the local MPS elements and the modelix Nodes
+        val branch = replicatedModel.getBranch()
+        runBlocking(dispatcher) {
```

Comment:
I am still not convinced that parallazing the traversal is worth the implementation complexity.
I would only consider parallelizing after measuring a bottleneck.
In the case of `MpsToModelixMapInitializerVisitor` everything is probably synced again because of the Read-lock `repository.runReadAction`.

Reply:
Read-actions can run in parallel